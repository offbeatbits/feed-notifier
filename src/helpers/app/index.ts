export * from "./isItForcedUpdate";
export * from "./isItInitialization";
