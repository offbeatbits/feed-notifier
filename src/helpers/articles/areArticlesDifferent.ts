import { logger } from "../../lib/logger";
import { ArticleObject } from "../../types";

const areStringsDifferent = (
  someString: string | null | undefined,
  otherString: string | null | undefined
) => {
  if (someString && otherString) {
    return someString.localeCompare(otherString) !== 0;
  }
  return false;
};

const areDatesDifferent = (
  oldDate: string | null | undefined,
  newDate: string | null | undefined
) => {
  if (oldDate && newDate) {
    return (
      new Date(oldDate!).getTime() !== new Date(newDate).getTime() ||
      Boolean(oldDate)
    );
  }
  return false;
};

interface ArticlesToCompare {
  oldArticle: ArticleObject | null | undefined;
  newArticle: ArticleObject | null | undefined;
  skip?: boolean;
}

export const areArticlesDifferent = ({
  oldArticle,
  newArticle,
  skip,
}: ArticlesToCompare) => {
  if (skip) {
    logger.verbose("Skipping article comparison due to 'skipCache' flag...");
    return true;
  }

  logger.verbose("Comparing articles...");

  const titleIsDifferent = areStringsDifferent(
    oldArticle?.title,
    newArticle?.title
  );

  logger.verbose(`Article title is different: ${titleIsDifferent}`);

  const urlIsDifferent = areStringsDifferent(oldArticle?.url, newArticle?.url);

  logger.verbose(`Article URL is different: ${urlIsDifferent}`);

  const articleDateIsLater = areDatesDifferent(
    oldArticle?.date,
    newArticle?.date
  );

  logger.verbose(`Article date is later: ${articleDateIsLater}`);

  const feedUpdateDateIsLater = areDatesDifferent(
    oldArticle?.feedUpdateDate,
    newArticle?.feedUpdateDate
  );

  logger.verbose(`Feed build date is later: ${feedUpdateDateIsLater}`);

  const result =
    titleIsDifferent &&
    urlIsDifferent &&
    articleDateIsLater &&
    feedUpdateDateIsLater;

  logger.verbose(`Articles are different: ${result}`);

  return result;
};
