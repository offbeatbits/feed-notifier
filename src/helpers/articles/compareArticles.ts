import { logger } from '../../lib/logger';
import { ArticleObject } from '../../types';

const compareStrings = (
  someString: string | null | undefined,
  otherString: string | null | undefined,
) => {
  if (someString && otherString) {
    return someString.localeCompare(otherString) === 0;
  }
  return false;
};

const compareDates = (
  oldDate: string | null | undefined,
  newDate: string | null | undefined,
) => {
  if (oldDate && newDate) {
    return new Date(oldDate!).getTime() < new Date(newDate).getTime()
      || Boolean(oldDate);
  }
  return false;
};

interface ArticlesToCompare {
  oldArticle: ArticleObject | null | undefined;
  newArticle: ArticleObject | null | undefined;
  skip?: boolean;
}

export const compareArticles = ({
  oldArticle,
  newArticle,
  skip,
}: ArticlesToCompare) => {
  if (skip) {
    logger.verbose("Skipping article comparison due to 'skipCache' flag...");
    return true;
  }

  logger.verbose('Comparing articles...');

  const titleIsDifferent = compareStrings(
    oldArticle?.title,
    newArticle?.title,
  );

  logger.verbose(`Article title is different: ${titleIsDifferent}`);

  const urlIsDifferent = compareStrings(
    oldArticle?.url,
    newArticle?.url,
  );

  logger.verbose(`Article URL is different: ${urlIsDifferent}`);

  const articleDateIsLater = compareDates(
    oldArticle?.date,
    newArticle?.date,
  );

  logger.verbose(`Article date is later: ${articleDateIsLater}`);

  const feedUpdateDateIsLater = compareDates(
    oldArticle?.feedUpdateDate,
    newArticle?.feedUpdateDate,
  );

  logger.verbose(`Feed build date is later: ${feedUpdateDateIsLater}`);

  const result = titleIsDifferent
    && urlIsDifferent
    && articleDateIsLater
    && feedUpdateDateIsLater;

  logger.verbose(`Articles are different: ${result}`);

  return result;
};
