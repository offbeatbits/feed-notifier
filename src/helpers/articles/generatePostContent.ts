import { logger } from "../../lib/logger";
import { ArticleObject } from "../../types.d";

export const generatePostContent = (article: ArticleObject) => {
  const content = `${article.title} ${article.url}`;
  logger.verbose(`Status content: ${content}`);
  return content;
};
