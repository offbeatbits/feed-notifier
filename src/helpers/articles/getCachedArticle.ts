import { logger } from "../../lib/logger";
// eslint-disable-next-line @typescript-eslint/no-redeclare
import { Cache } from "../../lib/cache";
import { app } from "../../config";

import { ArticleObject } from "../../types.d";

export const getCachedArticle = () => {
  logger.verbose("Getting cached article...");
  const article = Cache.get(app.cacheKey);

  if (article) {
    const parsedArticle = JSON.parse(article) as ArticleObject;

    const { feedUpdateDate, date, title, url } = parsedArticle;

    logger.verbose(`Cached feed update date: ${feedUpdateDate}`);
    logger.verbose(`Cached article date: ${date}`);
    logger.verbose(`Cached article title: ${title}`);
    logger.verbose(`Cached article URL: ${url}`);

    return parsedArticle;
  }
  return null;
};
