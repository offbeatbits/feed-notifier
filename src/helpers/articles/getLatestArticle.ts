import { logger } from '../../lib/logger';
import { fetchFeed } from '../feed';
import { ArticleObject } from '../../types';

export const getLatestArticle = async (feedUrl: string) => {
  logger.verbose('Getting latest article...');
  try {
    const feed = await fetchFeed(feedUrl);

    const article = {
      feedUpdateDate: feed.getFeedUpdateDate() || feed.getLatestPostDate(),
      date: feed.getLatestPostDate(),
      title: feed.getLatestPostTitle(),
      url: feed.getLatestPostUrl(),
    } as ArticleObject;

    logger.verbose(`Feed update date: ${article.feedUpdateDate}`);
    logger.verbose(`Latest article date: ${article.date}`);
    logger.verbose(`Latest article title: ${article.title}`);
    logger.verbose(`Latest article URL: ${article.url}`);

    return article;
  } catch {
    logger.verbose('Something went wrong while fetching feed, returning null!');
    return null;
  }
};
