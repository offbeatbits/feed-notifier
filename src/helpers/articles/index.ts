export * from './compareArticles';
export * from './generatePostContent';
export * from './getCachedArticle';
export * from './getLatestArticle';
export * from './updateArticleCache';
