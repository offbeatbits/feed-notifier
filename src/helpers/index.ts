export * from './createService';
export * from './getChannelArray';
