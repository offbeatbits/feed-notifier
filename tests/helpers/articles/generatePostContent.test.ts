import { generatePostContent } from '../../../src/helpers/articles';

describe('compareArticles', () => {
  const article = {
    feedUpdateDate: '2021-06-04T11:20:42Z',
    date: '2021-06-04T11:20:42Z',
    title: 'sample_title_1',
    url: 'sample_url_1',
  };

  it('should generate status post content for a given article', () => {
    expect(generatePostContent(article)).toMatchSnapshot();
  });
});
