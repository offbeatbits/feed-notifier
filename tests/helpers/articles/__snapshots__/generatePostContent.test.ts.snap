// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compareArticles should generate status post content for a given article 1`] = `"sample_title_1 sample_url_1"`;
